sklearn/metrics/_dist_metrics.pyx

```diff
@@ -554,27 +546,49 @@ cdef class MinkowskiDistance(DistanceMetric):
     r"""Minkowski Distance
 
     .. math::
-       D(x, y) = [\sum_i |x_i - y_i|^p] ^ (1/p)
+       D(x, y) = [\sum_i w_i *|x_i - y_i|^p] ^ (1/p)
+
+    Parameters
+    ----------
+    p : int
+        The order of the norm of the difference :math:`{||u-v||}_p`.
+    w : (N,) array-like (optional)
+        The weight vector.
 
     Minkowski Distance requires p >= 1 and finite. For p = infinity,
     use ChebyshevDistance.
     Note that for p=1, ManhattanDistance is more efficient, and for
     p=2, EuclideanDistance is more efficient.
     """
-    def __init__(self, p):
+    def __init__(self, p, w=None):
         if p < 1:
             raise ValueError("p must be greater than 1")
         elif np.isinf(p):
             raise ValueError("MinkowskiDistance requires finite p. "
                              "For p=inf, use ChebyshevDistance.")
+        elif w is not None and any(w_i < 0 for w_i in w):
```

Comment:
To avoid the scalar operations in the genexp, we could call `np.asarray(w, dtype=DTYPE)` ahead of time (if w is not None) and then:

And then check for `(w < 0).any()` which can be much more efficient for large arrays:

```python
In [1]: import numpy as np

In [2]: w = np.random.randn(int(1e5)) + 5

In [3]: %timeit any(w_i < 0 for w_i in w)
12.1 ms ± 54.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [4]: %timeit (w < 0).any()
43.9 µs ± 130 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```